Allow probing for /dev/fb cards other than the traditional sbus devices.

not suitable for upstream

--- xorg-server-21.1.20/hw/xfree86/common/xf86sbusBus.c
+++ xorg-server-21.1.20/hw/xfree86/common/xf86sbusBus.c
@@ -40,6 +40,10 @@ #include "xf86Bus.h"
 #include "xf86sbusBus.h"
 #include "xf86Sbus.h"
 
+#ifdef sun
+# include <sys/visual_io.h>
+#endif
+
 Bool sbusSlotClaimed = FALSE;
 
 static int xf86nSbusInfo;
@@ -231,8 +235,26 @@ xf86ErrorF(" at %s", promPath);
                 free(promPath);
             }
         }
-        else
+        else 
+#ifdef sun
+        {
+            if (psdp->devId == SBUS_DEVICE_UNK) {
+                int fbfd;
+
+                fbfd = open(psdp->device, O_RDONLY, 0);
+                if (fbfd >= 0) {
+                    struct vis_identifier   visid;
+                    if (ioctl(fbfd, VIS_GETIDENTIFIER, &visid) >= 0) {
+                        psdp->descr = strdup(visid.name);
+                    }
+                    close(fbfd);
+                }
+            }
+            xf86Msg(X_PROBED, "%s: %s", psdp->device, psdp->descr);
+        }
+#else
             xf86Msg(X_PROBED, "SBUS: %s", psdp->descr);
+#endif
         xf86ErrorF("\n");
     }
     if (useProm)
--- xorg-server-21.1.20/hw/xfree86/common/xf86sbusBus.h
+++ xorg-server-21.1.20/hw/xfree86/common/xf86sbusBus.h
@@ -39,6 +39,7 @@ #define SBUS_DEVICE_TCX		0x000a
 #define SBUS_DEVICE_FFB		0x000b
 #define SBUS_DEVICE_GT		0x000c
 #define SBUS_DEVICE_MGX		0x000d
+#define SBUS_DEVICE_UNK		0x000e
 
 typedef struct sbus_prom_node {
     int node;
--- xorg-server-21.1.20/hw/xfree86/os-support/bus/Sbus.c
+++ xorg-server-21.1.20/hw/xfree86/os-support/bus/Sbus.c
@@ -71,6 +71,7 @@ {SBUS_DEVICE_LEO, FBTYPE_SUNLEO, "leo",
     {SBUS_DEVICE_TCX, FBTYPE_TCXCOLOR, "tcx", "suntcx", "Sun TCX"},
     {SBUS_DEVICE_FFB, FBTYPE_CREATOR, "ffb", "sunffb", "Sun FFB"},
     {SBUS_DEVICE_FFB, FBTYPE_CREATOR, "afb", "sunffb", "Sun Elite3D"},
+    {SBUS_DEVICE_UNK, FBTYPE_LASTPLUSONE + 1, "unknown", "unknown", "Unknown FB"},
     {0, 0, NULL}
 };
 
